Automattic/WP-Job-Manager
# Register post meta and show default fields in REST API

Fixes #1697
Fixes #1591 

### Changes proposed in this Pull Request:

- Refactors the way job listing fields are registered in WP admin. *This must/should be backwards compatible*
- Registers `job_listing` post meta and enables `show_in_rest`. 
- Moves sanitization of post meta higher up into `add_post_meta`/`update_post_meta`. There is a new `sanitize_callback` attribute on field settings. There is a default sanitization for core fields that mimics previous behavior. Meta values coming from the frontend input are getting double-sanitized in some cases, but because they could be passed off to other areas (email notifications, plugins, etc), probably best to keep it that way.
- Moves auth for editing to callbacks as well in `auth_edit_callback` attribute on field settings. Right now these don't do much because the CPT has tighter permissions in REST API.
- Introduces `auth_view_callback` which defaults to previous behavior for WP admin. The only exception is `job_expires`. That field is hidden unless you're an admin or an employer viewing their own jobs.

### Fields Now Public by REST API
- `_job_location`
- `_application`
- `_company_name`
- `_company_website`
- `_company_tagline`
- `_company_twitter`
- `_company_video`
- `_filled`
- `_featured`

### What's not in this PR
- This is Phase 1 of working on WPJM's REST API. Right now, non-read events are very limited. Meta fields have looser `auth` rules, but aren't accessible because `job_listings` are only editable in REST API by admins. We'll eventually open this up a bit more so that employers can post/edit their own listings.
- I kept the logic for syncing job expiry/post status in `writepanels`. It is actually probably a bit broken in the block editor right now. I started moving that logic to higher level filters that could be shared among the REST API and front/backends, but it is going to require a bit of an over hall. 
- Google's location meta fields (`geolocation_*`). These should probably be added eventually, but in a different PR (possibly read-only). This would be the first example of `'show_in_admin' => false`.

### Known Backwards Breaking Changes
- `job_manager_job_listing_data_fields` no longer gets passed `$post_id`. It may not always be set, but if needed you could check the global `$post_id` (which is what was being passed before). For setting values manually, I'd use the new `job_manager_job_listing_wp_admin_fields`, which does get passed the `$post_id`.

### Testing Instructions
- First, make sure no behavior changes when creating/modifying job listings in WP admin in both the block and classic editors. Some fields might have a bit stricter sanitization, but shouldn't break standard usage.
- Test with themes / plugins that add custom fields to the backend editor. These should all work with this PR.
- `GET http://{$HOST}/wp-json/wp/v2/job-listings/` should return listings and their public meta fields under `meta`.
- `GET http://{$HOST}/wp-json/wp/v2/job-listings/{$ID}` should return a specific listing and its public meta fields under `meta`.
- Authenticate as admin (I use [JWT Authentication for WP REST API](https://wordpress.org/plugins/jwt-authentication-for-wp-rest-api/)), but lots of browser+cookie based options to. Verify `_job_expires` is passed along in the above requests.
- `PUT http://{$HOST}/wp-json/wp/v2/job-listings/{$ID}` should allow editing of fields in `meta` (authenticated as admin).
- `POST http://{$HOST}/wp-json/wp/v2/job-listings/` should allow creating job listings as admin with meta fields.
- Try various bad strings to make sure things that should be sanitized are.
- Test the front end job submission form as a guest (make sure settings allow for that), new employer (start as logged out), existing employer (log in first), and admin.

### New Filters
- `job_manager_job_listing_wp_admin_fields` should be used to modify field values/placeholders in WP admin. Do not use to add or remove fields.
